samples/bluetooth/hci_usb/src/main.c

```diff
@@ -661,14 +661,26 @@ DEVICE_INIT(btusb, "btusb", &btusb_init,
 
 static int try_write(u8_t ep, struct net_buf *buf)
 {
+	u8_t *data = buf->data;
+	u16_t len = buf->len;
+
 	while (1) {
-		int ret = usb_write(ep, buf->data, buf->len, NULL);
+		int ret;
+		u32_t wrote;
+
+		ret = usb_write(ep, data, len, &wrote);
 
 		switch (ret) {
 		case -EAGAIN:
+			k_yield();
 			break;
 		/* TODO: Handle other error codes */
 		default:
```

Comment:
Should you handle other error codes? BTW, could you use similar function try_write() in netusb.c

Reply:
Apart from EAGAIN, we can possibly get EIO, EINVAL, ETIMEDOUT from the current USB drivers in the case of errors.
In the case of other errors I should return instead of checking the wrote value. It should be done only when return value from usb_write is 0.

Also, I believe this patch is only required until usb_transfer() APIs are available in the device stack which would be the correct way to fix this. So for now, I'll handle the length check only if return value is zero. That will be similar to the try_write() implementation in netusb.c

Reply:
One other thing from netusb.c is that they try only for 10 times without yielding. I believe that's going to fail sometimes with certain drivers like the upcoming nRF52840 USB driver. Keeping it an arbitrary value will fail one or the other time. So for now, it's going to be a forever loop if -EAGAIN is returned all the time. If the driver is implemented properly this shouldn't happen though.